packages/notebook-extension/src/index.ts

```diff
@@ -763,6 +766,58 @@ function addCommands(app: JupyterLab, services: ServiceManager, tracker: Noteboo
     },
     isEnabled
   });
+  commands.addCommand(CommandIDs.runInConsole, {
+    label: 'Run Selected Text or Current Line in Console',
+    execute: args => {
+      // Default to not activating the notebook (thereby putting the notebook
+      // into command mode)
+      const current = getCurrent({activate: false, ...args});
+
+      if (current) {
+        const { context, content } = current;
+
+        let cell = content.activeCell;
+        let path = context.path;
+        // ignore action in non-code cell
+        if (!cell || cell.model.type !== 'code') {
+            return Promise.resolve(void 0);
+        }
+
+        let code = '';
+        const editor = cell.editor;
+        const selection = editor.getSelection();
+        const { start, end } = selection;
+        let selected = start.column !== end.column || start.line !== end.line;
+
+        if (selected) {
+          // Get the selected code from the editor.
+          const start = editor.getOffsetAt(selection.start);
+          const end = editor.getOffsetAt(selection.end);
+          code = editor.model.value.text.substring(start, end);
+        } else {
+          // no selection, submit whole line and advance
+          code = editor.getLine(selection.start.line);
+          const cursor = editor.getCursorPosition();
+          if (cursor.line + 1 !== editor.lineCount) {
+            editor.setCursorPosition({ line: cursor.line + 1, column: cursor.column });
```

Comment:
This should probably use the same logic we use in cursor-down, which goes to the next cell if you are at the bottom of a line.

Reply:
This "going to the next-cell" behavior can be convenient from time to time but it seems to break the "debugging this cell" mentality when I use this command. I mean, my work pattern is usually 

1. Execute the cell line by line using this command, revise the code if needed. Here I sometime press the shortcut repeatedly to observe if there is any error message.

2. When the cursor moves to the end of the cell (and the shortcut no longer works), I use `C-Enter` or `S-Enter` to execute the cell to generate proper output because the current output of the cell might not match the cell content after step 1.

Jumping to the next cell can make the current cell disappear, which can be irritating if the line just executed has some problem and need to be revised.

Reply:
What was surprising to me was that I executed per line until I got to the last line, then it kept executing the last line over and over again.

It was also a bit surprising to me that there was no natural behavior for executing the command when the notebook was in command mode. I think I would have expected it to execute the cell as a whole.

Reply:
We can open a ticket for discussion on proper end of cell and command mode behavior after merging so that people can try it in 0.33.